### PR TITLE
Creation of struct and functions to allow for temporary stat changes

### PR DESCRIPTION
**Pull request by jintae412**
Created support for temporary stat effects through the creation of a double linked list. The double linked list holds all of the stat changes and decrements the turns left by 1 each turn. Once the turns left is 0, it removes the item from the list. Basic tests have been written. Note: The function to add a node from an item does not yet add a turn count, because we have not yet changed other existing code to integrate this new functionality.

### CODE REVIEW

**Review by sonnykaiyuenlee — APPROVED** [review]
seems like you added a lot of good functionality to keep track of stat changes. It all makes sense to me, good job!

File: src/battle/src/battle_state.c [hunk 1]
```diff
@@ -153,3 +153,141 @@ int battle_free(battle_t *b)
 
     return SUCCESS;
 }
+
+/* See battle_state.h */
+stat_changes_t *stat_changes_new(){
```
Comment (firatciftci) [inline src/battle/src/battle_state.c:158]: A very minor stylistic thing: Make sure that the is a space between `)` and `{` here (and also in all of the other function declarations and/or if/for/while loop statements where applicable).

File: src/battle/src/battle_state.c [hunk 2]
```diff
@@ -153,3 +153,141 @@ int battle_free(battle_t *b)
...
+    stat_changes_t* sc;
+    int rc;
+
+    sc = calloc(1, sizeof(stat_changes_t));
+
+    if(sc == NULL)
```
Comment (firatciftci) [inline src/battle/src/battle_state.c:164]: Similar minor stylistic point: Make sure there is a single space between `if` or `for` or `while` and `(...` (here and wherever applicable in the code).

File: src/battle/src/battle_state.c [hunk 3]
```diff
@@ -153,3 +153,141 @@ int battle_free(battle_t *b)
...
+
+    return SUCCESS;
+}
+
+/* See battle_state.h */
+int stat_changes_turn_increment(stat_changes_t *sc, combatant_t *c){
+    stat_changes_t *current = sc->next;
+    stat_changes_t *remove = sc->next;
+
+    while(current != NULL){
```
Comment (firatciftci) [inline src/battle/src/battle_state.c:264]: For `for`, `while`, and `if` statements, the style must be like the following (there must be spaces around the statement bloc):

`while (...) {`

File: src/battle/src/battle_state.c [hunk 4]
```diff
@@ -153,3 +153,141 @@ int battle_free(battle_t *b)
...
+        current->turns_left -= 1;
+        
+        
+        if(current->turns_left == 0){
+            remove = current;
+            current = current->next;
+
+            stat_changes_undo(remove, c);
+            stat_changes_remove_node(remove);
+        }else{
```
Comment (firatciftci) [inline src/battle/src/battle_state.c:274]: Spaces around `else` here as well.

**Review by firatciftci — CHANGES_REQUESTED** [review]
I have requested minor stylistic changes, so once they are taken care of everything should be good to go (make sure to check the full extent of the files that you are working on and change the items that I mention below everywhere you see fit).

**Comment by jintae412** [comment]
I believe that all of the requested changes have been made. Please let me know if I missed anything or if there are any other changes that I should make.

**Review by firatciftci — APPROVED** [review]
Everything looks good to me, good work!

**Comment by jintae412** [comment]
Summary of PR: In this PR a new double linked list struct was created to handle temporary stat effects for the player (such as boosting attack by 30 for 3 turns). In addition, functions were made to facilitate this such as decrementing the turn count of all effects inside a given instance of this struct and removing the stat change if the turns it had left becomes 0, mechanically undoing the effects of the stat change while doing so. This code has not been integrated in our current code however, in that starting a new battle will not create an instance of this struct.

**Comment by firatciftci** [comment]
Issue Score: ✔️++

Comments:
Great job!

**Pull request merged**